lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   AutoDiffParameterIndices *checkedWrtParamIndices =
       attr->getParameterIndices();
 
+  // Predicate checking if a type has associated tangent and cotangent spaces.
+  auto hasAssociatedSpaces = [&](Type type) -> bool {
+    return (bool)type->getAutoDiffAssociatedVectorSpace(
+               AutoDiffAssociatedVectorSpaceKind::Tangent, lookupConformance) &&
+           (bool)type->getAutoDiffAssociatedVectorSpace(
+               AutoDiffAssociatedVectorSpaceKind::Cotangent, lookupConformance);
+  };
+
   // If checked wrt param indices are not specified, compute them using parsed
   // wrt param indices.
   if (!checkedWrtParamIndices) {
     AutoDiffParameterIndicesBuilder autoDiffParameterIndicesBuilder(
         originalFnTy);
     if (parsedWrtParams.empty()) {
+      // This is odd because Self is placed at the end (Args..., Self) due
+      // to existing currying logic. Unfortunately we want to talk about the
+      // underlying method which actually is (Args..., Self), but this function
+      // is (Self) -> (Args...) -> R, but we don't want to talk about normal
+      // curried functions.
+      // TODO: Clean all this up.
+      SmallVector<Type, 4> allWRTParamTypes;
       if (original->isStatic() || isa<ConstructorDecl>(original)) {
         auto *methodTy =
             original->getMethodInterfaceType()->castTo<AnyFunctionType>();
-        autoDiffParameterIndicesBuilder
-            .setParameters(0, methodTy->getNumParams());
+        if (methodTy->getResult()->getAs<AnyFunctionType>() != nullptr) {
+          TC.diagnose(attr->getLocation(),
+                      diag::differentiable_attr_no_currying);
+          return;
+        }
+        for (auto &param : methodTy->getParams()) {
+          allWRTParamTypes.push_back(param.getPlainType());
+        }
       } else {
-        autoDiffParameterIndicesBuilder.setAllParameters();
+        if (auto *fnTy = originalFnTy->getResult()->getAs<AnyFunctionType>()) {
```

Comment:
I wonder if you can reduce duplication with the `if (original->isStatic() || isa<ConstructorDecl>(original))` code path? I think the following may be equivalent:
```c++
auto *methodTy = original->getMethodInterfaceType()->castTo<AnyFunctionType>();
auto *fnTy = originalFnTy->getResult()->getAs<AnyFunctionType>();
```

If you could reduce both branches into one, that would be great!
Maybe sth like:
```cpp
auto *methodTy = originalResultTy->getAs<AnyFunctionType>();
// Dedupe diagnostic.
if (methodTy->getResult()->is<AnyFunctionType>() || !isInstanceMethod || ...) {
  TC.diagnose(attr->getLocation(), diag::differentiable_attr_no_currying);
  return;
}
// Dedupe recording parameter types.
for (auto &param : methodTy->getParams())
  allWrtParamTypes.push_back(param.getPlainType());
// ...
```